### PR TITLE
Add methods to PreTrainedModel to use PyTorch's BetterTransformer

### PR DESCRIPTION
**Pull request by fxmarty**
As per title.

Should be merged only on the next Optimum release that will include https://github.com/huggingface/optimum/pull/676

## Before submitting
Tests are still to be done.

## Who can review?

@younesbelkada @sgugger

### CODE REVIEW

**Comment by younesbelkada** [comment]
as a side note, since in the previous `optimum` versions the `save_pretrained` and `push_to_hub` methods [are not blocked](https://github.com/huggingface/optimum/blob/18e73f3ba4be33071f53650824fe625d3018af40/optimum/bettertransformer/transformation.py#L236), I propose to explicitly block them for transformed models in this PR and/or force users to use a certain version of `optimum`.

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -2955,6 +2956,35 @@ def register_for_auto_class(cls, auto_class="AutoModel"):
 
         cls._auto_class = auto_class
 
+    def to_bettertransformer(self) -> "PreTrainedModel":
+        """
+        Converts the model to use PyTorch's BetterTransformer through the [Optimum
+        integration](https://huggingface.co/docs/optimum/bettertransformer/overview).
+
+        Returns:
+            `PreTrainedModel`: The model converted to BetterTransformer.
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:2965]:
```suggestion
            [`PreTrainedModel`]: The model converted to BetterTransformer.
```
Is it still going to be the same type of model?

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -2955,6 +2956,35 @@ def register_for_auto_class(cls, auto_class="AutoModel"):
...
+        """
+        if not is_optimum_available():
+            raise ImportError("The package `optimum` is required to use BetterTransformer.")
+
+        from optimum.bettertransformer import BetterTransformer
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:2970]: I think a version check on optimum with a clear error message would be good? Also can the transform be applied twice? If not there should be a check and a clear error message as well.

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -2955,6 +2956,35 @@ def register_for_auto_class(cls, auto_class="AutoModel"):
...
+
+        return BetterTransformer.transform(self)
+
+    def reverse_bettertransformer(self):
+        """
+        Reverts the transformation from `to_bettertransformer` so that the original modeling is used.
+
+        Returns:
+            `PreTrainedModel`: The model converted back to the original modeling.
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:2979]:
```suggestion
            [`PreTrainedModel`]: The model converted back to the original modeling.
```

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -2955,6 +2956,35 @@ def register_for_auto_class(cls, auto_class="AutoModel"):
...
+        """
+        if not is_optimum_available():
+            raise ImportError("The package `optimum` is required to use BetterTransformer.")
+
+        from optimum.bettertransformer import BetterTransformer
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:2984]: Same comments here.

**Review by sgugger — COMMENTED** [review]
Thanks for adding those! I'd also make sure to document the methods and add something in the optimization guides we have :-)

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (fxmarty) [reply src/transformers/modeling_utils.py:2970]: > Also can the transform be applied twice? If not there should be a check and a clear error message as well.

I think there's currently no check for this @younesbelkada . I will add it on the Optimum side. So as to keep the transformers side as lightweight as possible.

**Comment by fxmarty** [comment]
Should be ready @sgugger , the documentation has been extended in https://moon-ci-docs.huggingface.co/docs/transformers/pr_21259/en/perf_infer_gpu_one .

Let me know if I should add a test - in which case optimum should be added in the setup.py, I guess.

**Comment by younesbelkada** [comment]
@fxmarty there should be no need to add `optimum` in `setup.py`, we can do something similar than `bitsandbytes` and add `optimum` in the Dockerfile of the Docker image that will run the slow tests: https://github.com/huggingface/transformers/blob/0db5d911fc94604f9568b4b212e005ec4600d157/docker/transformers-all-latest-gpu/Dockerfile#L52 
I very much agree that we should add tests, especially to test `accelerate` compatibility, happy to help you on this, let me know if you need help

**Comment by fxmarty** [comment]
> especially to test accelerate compatibility

Isn't this already tested on Optimum side?

**Comment by younesbelkada** [comment]
> Isn't this already tested on Optimum side? 
I don't think so since the tests[ are run on GPU](https://github.com/huggingface/optimum/blob/40a01b3c883ca3c092a4493d3f5ca524ed3109ab/tests/bettertransformer/test_bettertransformer_encoder.py#L186 ): therefore not run on the CI runners (but not sue if there are tested somewhere else) - I just asked individually to each contributor to run the `accelerate` test locally on their GPU before merging.

**Comment by fxmarty** [comment]
There are tests on the daily basis on GPU in Optimum, for example https://github.com/huggingface/optimum/blob/main/.github/workflows/test_onnxruntime_train.yml and https://github.com/huggingface/optimum/blob/main/.github/workflows/test_onnxruntime_gpu.yml

In my opinion, thorough tests should be added in Optimum, not Transformers. The test I was thinking of in Transformers was only an integration one to check that there's no error.

**Comment by younesbelkada** [comment]
There is an issue with `accelerate` loaded models and `transform` from BT, let's wait until this gets fixed before merging this PR

**Comment by fxmarty** [comment]
not stale

**Comment by younesbelkada** [comment]
Thanks for the headsup! 
Indeed we are working on fixing some bugs on `optimum` side that was introduced by one of my PRs (the revert-transform PR) before adding the `invert_transform` method
We can maybe merge this PR by keeping only `transform` method and blocking the `save_pretrained` & `push_to_hub` methods after transforming the model

**Comment by fxmarty** [comment]
> you should finish the work and have it merged sooner rather than later :-)

There is substantial work left in Optimum before this should be merged. Marking as draft for now!

**Comment by sgugger** [comment]
OK, so this won't be in the next release of Transformers (probably this week in preparation for PyTorch 2.0).

**Comment by LysandreJik** [comment]
Hey @fxmarty and @younesbelkada, are there standing PRs in `optimum` that need to be merged for this to proceed/anything we can help with to have this move forward? Thanks :)

**Comment by younesbelkada** [comment]
Hey @LysandreJik @sgugger 
@fxmarty recently managed to fix all issues related to decoder-based models integration in `optimum`! I believe that this PR could be re-opened, in my understanding we just need to add few tests and we should be good to go

**Comment by younesbelkada** [comment]
@sgugger @LysandreJik this is now ready for review!

File: src/transformers/__init__.py [hunk 5]
```diff
@@ -103,6 +103,7 @@
         "is_clearml_available",
         "is_comet_available",
         "is_neptune_available",
+        "is_optimum_available",
```
Comment (sgugger) [inline src/transformers/__init__.py:106]: This is laready in `import_utils`.

File: src/transformers/integrations.py [hunk 6]
```diff
@@ -98,6 +98,10 @@ def is_tensorboard_available():
     return importlib.util.find_spec("tensorboard") is not None or importlib.util.find_spec("tensorboardX") is not None
 
 
+def is_optimum_available():
+    return importlib.util.find_spec("optimum") is not None
```
Comment (sgugger) [inline src/transformers/integrations.py:102]: No need to redefine it it already exists in `import_utils`.

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -37,6 +37,7 @@
 from .deepspeed import deepspeed_config, is_deepspeed_zero3_enabled
 from .dynamic_module_utils import custom_object_save
 from .generation import GenerationConfig, GenerationMixin
+from .integrations import is_optimum_available
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:40]:
```suggestion
from .utils import is_optimum_available
```

File: src/transformers/testing_utils.py [hunk 8]
```diff
@@ -42,6 +42,7 @@
 from .integrations import (
     is_clearml_available,
     is_fairscale_available,
+    is_optimum_available,
```
Comment (sgugger) [inline src/transformers/testing_utils.py:45]: Import from utils.

**Review by sgugger — COMMENTED** [review]
Thanks! Just have one comment on the `is_optimum_available` function but the rest looks fine!

File: src/transformers/testing_utils.py (hunk 8 above)
Comment (younesbelkada) [reply src/transformers/testing_utils.py]: Fixed in [9c5e0d9](https://github.com/huggingface/transformers/pull/21259/commits/9c5e0d9308b5c4882805058023729fb6dd63cbe9)

File: src/transformers/integrations.py (hunk 6 above)
Comment (younesbelkada) [reply src/transformers/integrations.py]: Thanks, fixed in [9c5e0d9](https://github.com/huggingface/transformers/pull/21259/commits/9c5e0d9308b5c4882805058023729fb6dd63cbe9)

File: src/transformers/__init__.py (hunk 5 above)
Comment (younesbelkada) [reply src/transformers/__init__.py]: Fixed in [9c5e0d9](https://github.com/huggingface/transformers/pull/21259/commits/9c5e0d9308b5c4882805058023729fb6dd63cbe9)

**Review by sgugger — APPROVED** [review]
Thanks!

File: docs/source/en/perf_infer_gpu_one.mdx [hunk 9]
```diff
@@ -11,11 +11,28 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
 
-## `BetterTransformer` for faster inference
+## Better Transformer: PyTorch-native transformer fastpath
 
-We have recently integrated `BetterTransformer` for faster inference on GPU for text, image and audio models. Check the documentation about this integration [here](https://huggingface.co/docs/optimum/bettertransformer/overview) for more details.
+PyTorch-native [`nn.MultiHeadAttention`](https://pytorch.org/blog/a-better-transformer-for-fast-transformer-encoder-inference/) attention fastpath, called BetterTransformer, can be used with Transformers through the integration in the [🤗 Optimum library](https://huggingface.co/docs/optimum/bettertransformer/overview).
+
+PyTorch's attention fastpath allows to speed up inference through kernel fusions and the use of [nested tensors](https://pytorch.org/docs/stable/nested.html). Detailed benchmarks can be found in [this blog post](https://medium.com/pytorch/bettertransformer-out-of-the-box-performance-for-huggingface-transformers-3fbe27d50ab2).
+
+After installing the `optimum` package, to use Better Transformer during inference, the relevant internal modules are replaced by calling [`~PreTrainedModel.to_bettertransformer`]:
```
Comment (michaelbenayoun) [inline docs/source/en/perf_infer_gpu_one.mdx:22]:
```suggestion
After installing the [`optimum`](https://github.com/huggingface/optimum) package, to use Better Transformer during inference, the relevant internal modules are replaced by calling [`~PreTrainedModel.to_bettertransformer`]:
```

File: src/transformers/modeling_utils.py [hunk 10]
```diff
@@ -3310,6 +3311,56 @@ def register_for_auto_class(cls, auto_class="AutoModel"):
...
+        """
+        if not is_optimum_available():
+            raise ImportError("The package `optimum` is required to use Better Transformer.")
+
+        from optimum.version import __version__ as optimum_version
+
+        if version.parse(optimum_version) < version.parse("1.7.0"):
+            raise ImportError(
+                f"Please install optimum>=1.7.0 to use Better Transformer. The version {optimum_version} was found."
+            )
```
Comment (michaelbenayoun) [inline src/transformers/modeling_utils.py:3336]: Maybe factor all of this into a `is_bettertransformer_available`?

File: tests/bettertransformer/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+
```
Comment (michaelbenayoun) [inline tests/bettertransformer/__init__.py:1]: Is it wanted?

**Review by michaelbenayoun — APPROVED** [review]
Left a few nits.
LGTM!

File: src/transformers/modeling_utils.py (hunk 10 above)
Comment (younesbelkada) [reply src/transformers/modeling_utils.py:3336]: Hmm I would say this is too specific, maybe let's keep it as it is

**Pull request merged**